File: code/preprocTimingAndPitch.R
# readAloud-valence-dataset Timingn and Pitch Preprocessing
# Author: Jessica M. Alexander
# Last Updated: 2022-04-14

### SECTION 1: SETTING UP
library(readxl)

#set up date for output file naming
today <- Sys.Date()
today <- format(today, "%Y%m%d")

#set up directories for input/output data
#hpc
#passageChar_path <- '/home/data/NDClab/datasets/readAloud-valence-dataset/materials/readAloud-ldt/stimuli/readAloud/readAloud-stimuli_characteristics.xlsx'
#input_path <- '/home/data/NDClab/datasets/readAloud-valence-dataset/derivatives/preprocessed/timing-and-pitch/'
#out_path <- '/home/data/NDClab/datasets/readAloud-valence-dataset/derivatives/preprocessed/timing-and-pitch/'
#local
passageChar_path <- '/Users/jalexand/github/readAloud-valence-dataset/materials/readAloud-ldt/stimuli/readAloud/readAloud-stimuli_characteristics.xlsx'
input_path <- '/Users/jalexand/github/readAloud-valence-dataset/derivatives/preprocessed/timing-and-pitch/'
out_path <- '/Users/jalexand/github/readAloud-valence-dataset/derivatives/preprocessed/timing-and-pitch/'

#load passage characteristics data
passageChar <- read_excel(passageChar_path, sheet="passages")

#identify participant folders within input dir
sub_files <- list.files(input_path, pattern = "sub")

#create dataframes for storing output data and define output file names
#timing
timingSummaryDat <- data.frame(matrix(ncol=13, nrow=0))
colnames(timingSummaryDat) <- c("id",
                               "posFirstSpeed", "negFirstSpeed",
                               "posLastSpeed", "negLastSpeed",
                               "prePREswitchSpeed_pos2neg", "prePREswitchSpeed_neg2pos",
                               "preswitchSpeed_pos2neg", "preswitchSpeed_neg2pos",
                               "switchSpeed_pos2neg", "switchSpeed_neg2pos",
                               "postswitchSpeed_pos2neg", "postswitchSpeed_neg2pos")
timing_out_subjectLevel <- paste("timing_subject-level_summary_", today, ".csv", sep="", collapse=NULL)

timingTrialDat <- data.frame(matrix(ncol=9, nrow=0))
colnames(timingTrialDat) <- c("id",
                             "passage",
                             "switchType",
                             "firstSpeed",
                             "lastSpeed",
                             "prePREswitchSpeed",
                             "preswitchSpeed",
                             "switchSpeed",
                             "postswitchSpeed")
timing_out_trialLevel <- paste("timing_trial-level_summary_", today, ".csv", sep="", collapse=NULL)

#pitch
pitchSummaryDat <- data.frame(matrix(ncol=13, nrow=0))
colnames(pitchSummaryDat) <- c("id",
                               "pitchStartToSwitch_pos2neg", "pitchStartToSwitch_neg2pos",
                               "pitchSwitchToEnd_pos2neg", "pitchSwitchToEnd_pos2neg",
                               "pitchPrePREswitchGroup_pos2neg", "pitchPrePREswitchGroup_neg2pos",
                               "pitchPreswitchGroup_pos2neg", "pitchPreswitchGroup_neg2pos",
                               "pitchSwitchGroup_pos2neg", "pitchSwitchGroup_neg2pos",
                               "pitchPostswitchGroup_pos2neg", "pitchPostswitchGroup_neg2pos")
pitch_out_subjectLevel <- paste("pitch_subject-level_summary_", today, ".csv", sep="", collapse=NULL)

pitchTrialDat <- data.frame(matrix(ncol=9, nrow=0))
colnames(pitchTrialDat) <- c("id",
                              "passage",
                              "switchType",
                              "firstPitch",
                              "lastPitch",
                              "prePREswitchPitch",
                              "preswitchPitch",
                              "switchPitch",
                              "postswitchPitch")
pitch_out_trialLevel <- paste("pitch_trial-level_summary_", today, ".csv", sep="", collapse=NULL)


### SECTION 2: START PARTICIPANT LOOP AND CLEAN INCOMING DATAFRAME
#loop over participants (subfiles)
for(i in 1:length(sub_files)){
  
  #find the timing and pitch file for this participant
  timingPitch_file <- paste(input_path,sub_files[i], sep = "", collapse = NULL)
  print(paste("Woohoo! Processing ", sub_files[i], "!"))

  #read in the data for this participant, establish id, and add passage type (pos2neg or neg2pos)
  timingPitchDat <- read_excel(timingPitch_file, sheet = NULL)
  id <- strsplit(sub_files[i], "_")[[1]][1]
  timingPitchDat$switchType <- passageChar$switchType
  
  
  ### SECTION 3: SPLIT INTO DATAFRAMES BY SWITCH TYPE
  timingPitchDat_pos2neg <- timingPitchDat[timingPitchDat$switchType=="pos2neg",]
  timingPitchDat_neg2pos <- timingPitchDat[timingPitchDat$switchType=="neg2pos",]
  
  
  ### SECTION 4: CALCULATE PASSAGE HALF READING SPEEDS
  #first passage half
  timingPitchDat_pos2neg$posFirstTime <- timingPitchDat_pos2neg$switchWord - timingPitchDat_pos2neg$readStart
  posFirstLength <- NUMBER OF SYLLABLES HERE
  timingPitchDat_pos2neg$posFirstSpeed <- timingPitchDat_pos2neg$posFirstTime/posFirstLength
  posFirstSpeed <- mean(timingPitchDat_pos2neg$posFirstSpeed)
  
  timingPitchDat_neg2pos$negFirstTime <- timingPitchDat_neg2pos$switchWord - timingPitchDat_neg2pos$readStart
  negFirstLength <- NUMBER OF SYLLABLES HERE
  timingPitchDat_neg2pos$negFirstSpeed <- timingPitchDat_neg2pos$negFirstTime/negFirstLength
  negFirstSpeed <- mean(timingPitchDat_neg2pos$negFirstSpeed)
  
  #second passage half
  timingPitchDat_pos2neg$posLastTime <- timingPitchDat_pos2neg$readEnd - timingPitchDat_pos2neg$switchWord
  posLastLength <- NUMBER OF SYLLABLES HERE
  timingPitchDat_pos2neg$posLastSpeed <- timingPitchDat_pos2neg$posLastTime/posLastLength
  posLastSpeed <- mean(timingPitchDat_pos2neg$posLastSpeed)
  
  timingPitchDat_neg2pos$negLastTime <- timingPitchDat_neg2pos$readEnd - timingPitchDat_neg2pos$switchWord
  negLastLength <- NUMBER OF SYLLABLES HERE
  timingPitchDat_neg2pos$negLastSpeed <- timingPitchDat_neg2pos$negLastTime/negLastLength
  negLastSpeed <- mean(timingPitchDat_neg2pos$negLastSpeed)
  
  ### SECTION 5: CALCULATE PASSAGE HALF READING PITCH
  pitchStartToSwitch_pos2neg <- mean(timingPitchDat_pos2neg$pitchStartToSwitch)
  pitchStartToSwitch_neg2pos <- mean(timingPitchDat_neg2pos$pitchStartToSwitch)
  pitchSwitchToEnd_pos2neg <- mean(timingPitchDat_pos2neg$pitchSwitchToEnd)
  pitchSwitchToEnd_neg2pos <- mean(timingPitchDat_neg2pos$pitchSwitchToEnd)
  
  ### SECTION 6: CALCULATE SWITCH GROUP READING SPEED
  #pos2neg: prePREswitch group
  timingPitchDat_pos2neg$prePREswitchTime_pos2neg <- timingPitchDat_pos2neg$preswitchStart - timingPitchDat_pos2neg$prePREswitchStart
  prePREswitchLength_pos2neg <- NUMBER OF SYLLABLES HERE
  timingPitchDat_pos2neg$prePREswitchSpeed_pos2neg <- timingPitchDat_pos2neg$prePREswitchTime_pos2neg/prePREswitchLength_pos2neg
  prePREswitchSpeed_pos2neg <- mean(timingPitchDat_pos2neg$prePREswitchSpeed_pos2neg)
  
  #pos2neg: preswitch group
  timingPitchDat_pos2neg$preswitchTime_pos2neg <- timingPitchDat_pos2neg$switchStart - timingPitchDat_pos2neg$preswitchStart
  preswitchLength_pos2neg <- NUMBER OF SYLLABLES HERE
  timingPitchDat_pos2neg$preswitchSpeed_pos2neg <- timingPitchDat_pos2neg$preswitchTime_pos2neg/preswitchLength_pos2neg
  preswitchSpeed_pos2neg <- mean(timingPitchDat_pos2neg$preswitchSpeed_pos2neg)
  
  #pos2neg: switch group
  timingPitchDat_pos2neg$switchTime_pos2neg <- timingPitchDat_pos2neg$postswitchStart - timingPitchDat_pos2neg$switchStart
  switchLength_pos2neg <- NUMBER OF SYLLABLES HERE
  timingPitchDat_pos2neg$switchSpeed_pos2neg <- timingPitchDat_pos2neg$switchTime_pos2neg/switchLength_pos2neg
  switchSpeed_pos2neg <- mean(timingPitchDat_pos2neg$switchSpeed_pos2neg)
  
  #pos2neg: postswitch group
  timingPitchDat_pos2neg$postswitchTime_pos2neg <- timingPitchDat_pos2neg$postswitchEnd - timingPitchDat_pos2neg$postswitchStart
  postswitchLength_pos2neg <- NUMBER OF SYLLABLES HERE
  timingPitchDat_pos2neg$postswitchSpeed_pos2neg <- timingPitchDat_pos2neg$postswitchTime_pos2neg/postswitchLength_pos2neg
  postswitchSpeed_pos2neg <- mean(timingPitchDat_pos2neg$postswitchSpeed_pos2neg)
  
  #neg2pos: prePREswitch group
  timingPitchDat_neg2pos$prePREswitchTime_neg2pos <- timingPitchDat_neg2pos$preswitchStart - timingPitchDat_neg2pos$prePREswitchStart
  prePREswitchLength_neg2pos <- NUMBER OF SYLLABLES HERE
  timingPitchDat_neg2pos$prePREswitchSpeed_neg2pos <- timingPitchDat_neg2pos$prePREswitchTime_neg2pos/prePREswitchLength_neg2pos
  prePREswitchSpeed_neg2pos <- mean(timingPitchDat_neg2pos$prePREswitchSpeed_neg2pos)
  
  #neg2pos: preswitch group
  timingPitchDat_neg2pos$preswitchTime_neg2pos <- timingPitchDat_neg2pos$switchStart - timingPitchDat_neg2pos$preswitchStart
  preswitchLength_neg2pos <- NUMBER OF SYLLABLES HERE
  timingPitchDat_neg2pos$preswitchSpeed_neg2pos <- timingPitchDat_neg2pos$preswitchTime_neg2pos/preswitchLength_neg2pos
  preswitchSpeed_neg2pos <- mean(timingPitchDat_neg2pos$preswitchSpeed_neg2pos)
  
  #neg2pos: switch group
  timingPitchDat_neg2pos$switchTime_neg2pos <- timingPitchDat_neg2pos$postswitchStart - timingPitchDat_neg2pos$switchStart
  switchLength_neg2pos <- NUMBER OF SYLLABLES HERE
  timingPitchDat_neg2pos$switchSpeed_neg2pos <- timingPitchDat_neg2pos$switchTime_neg2pos/switchLength_neg2pos
  switchSpeed_neg2pos <- mean(timingPitchDat_neg2pos$switchSpeed_neg2pos)
  
  #neg2pos: postswitch group
  timingPitchDat_neg2pos$postswitchTime_neg2pos <- timingPitchDat_neg2pos$postswitchEnd - timingPitchDat_neg2pos$postswitchStart
  postswitchLength_neg2pos <- NUMBER OF SYLLABLES HERE
  timingPitchDat_neg2pos$postswitchSpeed_neg2pos <- timingPitchDat_neg2pos$postswitchTime_neg2pos/postswitchLength_neg2pos
  postswitchSpeed_neg2pos <- mean(timingPitchDat_neg2pos$postswitchSpeed_neg2pos)
  
  ### SECTION 7: CALCULATE SWITCH GROUP READING PITCH
  pitchPrePREswitchGroup_pos2neg <- mean(timingPitchDat_pos2neg$pitchPrePREswitchGroup)
  pitchPrePREswitchGroup_neg2pos <- mean(timingPitchDat_neg2pos$pitchPrePREswitchGroup)
  
  pitchPreswitchGroup_pos2neg <- mean(timingPitchDat_pos2neg$pitchPreswitchGroup)
  pitchPreswitchGroup_neg2pos <- mean(timingPitchDat_neg2pos$pitchPreswitchGroup)
  
  pitchSwitchGroup_pos2neg <- mean(timingPitchDat_pos2neg$pitchSwitchGroup)
  pitchSwitchGroup_neg2pos <- mean(timingPitchDat_neg2pos$pitchSwitchGroup)
  
  pitchPostswitchGroup_pos2neg <- mean(timingPitchDat_pos2neg$pitchPostswitchGroup)
  pitchPostswitchGroup_neg2pos <- mean(timingPitchDat_neg2pos$pitchPostswitchGroup)

  #store output data in summary matrices
  timingSummaryDat[nrow(timingSummaryDat) + 1,] <-c(id,
                                                    posFirstSpeed, negFirstSpeed,
                                                    posLastSpeed, negLastSpeed,
                                                    prePREswitchSpeed_pos2neg, prePREswitchSpeed_neg2pos,
                                                    preswitchSpeed_pos2neg, preswitchSpeed_neg2pos,
                                                    switchSpeed_pos2neg, switchSpeed_neg2pos,
                                                    postswitchSpeed_pos2neg, postswitchSpeed_neg2pos)

  pitchSummaryDat[nrow(pitchSummaryDat) +1,] <- c(id,
                                                  pitchStartToSwitch_pos2neg, pitchStartToSwitch_neg2pos,
                                                  pitchSwitchToEnd_pos2neg, pitchSwitchToEnd_pos2neg,
                                                  pitchPrePREswitchGroup_pos2neg, pitchPrePREswitchGroup_neg2pos,
                                                  pitchPreswitchGroup_pos2neg, pitchPreswitchGroup_neg2pos,
                                                  pitchSwitchGroup_pos2neg, pitchSwitchGroup_neg2pos,
                                                  pitchPostswitchGroup_pos2neg, pitchPostswitchGroup_neg2pos)
  
  
  
  for (j in 1:nrow(ldtDat)){
    trial_no <- j
    acc <- ldtDat$ldtResponse.corr[j]
    rt <- ldtDat$ldtResponse.rt[j]
    stimString <- ldtDat$ldtStim[j]
    wordType <- ldtDat$wordType[j]
    
    ldtTrialDat[nrow(ldtTrialDat) + 1,] <-c(id, trial_no, acc, rt, stimString, wordType)
    }
  
  
}



### SECTION 7: OUTPUT DATA
#write the extracted summary scores to CSV
write.csv(readAloudChallengeDat,paste(out_path,readAloud_out_passageLevel, sep = "", collapse = NULL), row.names=FALSE)
write.csv(readAloudSummaryDat,paste(out_path,readAloud_out_subjectLevel, sep = "", collapse = NULL), row.names=FALSE)
write.csv(ldtSummaryDat,paste(out_path,ldt_out_subjectLevel, sep = "", collapse = NULL), row.names=FALSE)
write.csv(ldtTrialDat,paste(out_path,ldt_out_trialLevel, sep = "", collapse = NULL), row.names=FALSE)
write.csv(dccsSummaryDat,paste(out_path,dccs_out_subjectLevel, sep = "", collapse = NULL), row.names=FALSE)
write.csv(dccsTrialDat,paste(out_path,dccs_out_trialLevel, sep = "", collapse = NULL), row.names=FALSE)

File: code/preprocTimingAndPitch.py
# readAloud-valence-dataset Timing and Pitch Preprocessing
# Author: Jessica M. Alexander
# Last Updated: 2022-09-06

### SECTION 1: SETTING UP
import numpy as np
import os
import pandas as pd
import parselmouth
import re
from datetime import date

today = date.today()
today_formatted = today.strftime("%Y-%m-%d")

#function to create working df for each subject and prefill with coded timestamps
def create_newdf(npdf):
  newdf = np.full((20,9), fill_value=np.NaN)
  
  for row in range(npdf.shape[0]):
    newdf[row, 0:3] = npdf[row, 1:4]
  
  return newdf

#function to extract data from praat using parselmouth
def extract_parselmouth(sub_id, pitch, df, newdf):
  for row in range(newdf.shape[0]):
    readStart = newdf[row,0]
    switchWord = newdf[row,1]
    readEnd = newdf[row,2]
    
    if np.isnan(switchWord):
      continue #cannot calculate any passage data if missing timestamp for switchWord
    if np.isnan(readStart) & np.isnan(readEnd):
      continue #cannot calculate any passage data if missing timestamp for both start and finish
    elif np.isnan(readEnd):
      newdf[row, 3] = switchWord - readStart #first half reading time
      newdf[row, 5] = parselmouth.praat.call(pitch, "Get mean", readStart, switchWord, "Hertz") #first half mean pitch
      newdf[row, 6] = parselmouth.praat.call(pitch, "Get standard deviation", readStart, switchWord, "Hertz") #first half pitch sd
    elif np.isnan(readStart):  
      newdf[row, 4] = readEnd - switchWord #second half reading time
      newdf[row, 7] = parselmouth.praat.call(pitch, "Get mean", switchWord, readEnd, "Hertz") #second half mean pitch
      newdf[row, 8] = parselmouth.praat.call(pitch, "Get standard deviation", switchWord, readEnd, "Hertz") #second half pitch sd
    else:
      newdf[row, 3] = switchWord - readStart #first half reading time
      newdf[row, 5] = parselmouth.praat.call(pitch, "Get mean", readStart, switchWord, "Hertz") #first half mean pitch
      newdf[row, 6] = parselmouth.praat.call(pitch, "Get standard deviation", readStart, switchWord, "Hertz") #first half pitch sd
      newdf[row, 4] = readEnd - switchWord #second half reading time
      newdf[row, 7] = parselmouth.praat.call(pitch, "Get mean", switchWord, readEnd, "Hertz") #second half mean pitch
      newdf[row, 8] = parselmouth.praat.call(pitch, "Get standard deviation", switchWord, readEnd, "Hertz") #second half pitch sd

  sublist = [sub_id] * newdf.shape[0]
  newdfplus = np.vstack((sublist, np.transpose(newdf)))
  dfout = np.transpose(np.vstack((newdfplus, df.iloc[:,0])))
  
  return dfout

#function to run each participant
def run_sub(sub, filelist, zoompath, timestamps, dfmain):
  filename = filelist[sub]
  coded_filepath = timestamps + filename
  subject = filename.split("_")[0]
  sub_id = int(subject.split("-")[1])
  subzoom = os.listdir(zoompath + subject) #list of all files in subject's zoom folder
  audiopath = zoompath + subject + "/" + subzoom[0]
  audio = parselmouth.Sound(audiopath) #feed .wav file to Parselmouth
  df = pd.read_excel(coded_filepath, header=0)
  npdf = df.to_numpy()
  pitch = audio.to_pitch_ac() #extract Parselmouth pitch object
  
  newdf = create_newdf(npdf) #create subject scaffold df
  
  dfout = extract_parselmouth(sub_id, pitch, df, newdf) #extract all sub data points from audio file with Parselmouth

  dfmain = np.vstack((dfmain, dfout)) #output array and append to dfmain
  
  return dfmain

### SECTION 2: FILE PATHS AND INITIALIZATION
#capture list of coded files
timestamps_ja = "/Users/jalexand/github/readAloud-valence-dataset/derivatives/preprocessed/valence-timing/coders/ja/"
timestamps_lg = "/Users/jalexand/github/readAloud-valence-dataset/derivatives/preprocessed/valence-timing/coders/lg/"
timestamps_mr = "/Users/jalexand/github/readAloud-valence-dataset/derivatives/preprocessed/valence-timing/coders/mr/"
zoompath = "/Users/jalexand/github/readAloud-valence-dataset/sourcedata/checked/zoom/"
outpath = "/Users/jalexand/github/readAloud-valence-dataset/derivatives/preprocessed/valence-timing/"

filelist_ja = os.listdir(timestamps_ja)
filelist_lg = os.listdir(timestamps_lg)
filelist_mr = os.listdir(timestamps_mr)

#initialize empty array
dfmain = np.empty((0,11))

### SECTION 3: RUN THREE PARTICIPANT LOOPS (one for each coder)
#loop over subjects to read in coded timestamps and extract reading times and pitch information
for sub in range(len(filelist_ja)):
  dfmain = run_sub(sub, filelist_ja, zoompath, timestamps_ja, dfmain)
  
for sub in range(len(filelist_lg)):
  dfmain = run_sub(sub, filelist_lg, zoompath, timestamps_lg, dfmain)
  
for sub in range(len(filelist_mr)):
  dfmain = run_sub(sub, filelist_mr, zoompath, timestamps_mr, dfmain)

### SECTION 4: OUTPUT DATA
columns = ['id', 'readStart','switchWord','readEnd', 'timeFirst', 'timeSecond', 'pitchMeanFirst', 'pitchSdFirst', 'pitchMeanSecond', 'pitchSdSecond', 'passage']
combodf = pd.DataFrame(dfmain, columns=columns)
combodf['id'] = combodf['id'].astype(int)
combodf.to_csv(outpath + "timingpitch_subject-by-passage_" + today_formatted + ".csv", index=False)
